### ISSUE TITLE
Confusing error messages when invariant type is involved

### ISSUE THREAD

**Issue by jonas-schievink** [rust-lang/rust#26804]
Playpen: http://is.gd/f1SPdn

```rust
#[derive(Default)]
struct S<'a> {
    c: RefCell<Option<Box<Sized + 'a>>>,    // Doesn't work
    //c: RefCell<Option<Box<&'a Sized>>>,   // Works
    //c: Option<Box<Sized + 'a>>,           // Also works
}

impl <'a> S<'a> {
    fn r(&'a self) -> &'a () {
        panic!();
    }
}

fn main() {
    let s = S::default();
    let r = s.r();
}
```

Gives this error:

```
<anon>:19:13: 19:14 error: `s` does not live long enough
<anon>:19     let r = s.r();
                      ^
<anon>:17:11: 20:2 note: reference must be valid for the block at 17:10...
<anon>:17 fn main() {
<anon>:18     let s = S::default();
<anon>:19     let r = s.r();
<anon>:20 }
<anon>:18:26: 20:2 note: ...but borrowed value is only valid for the block suffix following statement 0 at 18:25
<anon>:18     let s = S::default();
<anon>:19     let r = s.r();
<anon>:20 }
```

This is confusing, since the compiler seems to demand that `s` is valid in the line before it is declared (the only difference between the regions is that the first one includes the `fn main() {`). It is never mentioned that this is caused by `S` being invariant on `'a`, which disallows rustc to enlarge the lifetime to include the whole block.

I think rustc should give a better error is this case (maybe something like ``note: cannot extend lifetime of `s` since its type `S` is invariant with respect to `'a` ``).

**Comment by steveklabnik** [comment]
A totally different error, 

```
error[E0038]: the trait `std::marker::Sized` cannot be made into an object
 --> <anon>:6:5
  |
6 |     c: RefCell<Option<Box<Sized + 'a>>>,  // Doesn't work
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::marker::Sized` cannot be made into an object
  |
  = note: the trait cannot require that `Self : Sized`
```

in all three cases. However, if we introduce a new `Foo` trait rather than `Sized`, the same error happens:

https://is.gd/L91Hku

**Comment by steveklabnik** [comment]
Triage: same as my last comment

**Comment by estebank** [comment]
Current output:

```
error[E0597]: `s` does not live long enough
  --> src/main.rs:20:13
   |
20 |     let r = s.r();
   |             ^ borrowed value does not live long enough
21 | }
   | -
   | |
   | `s` dropped here while still borrowed
   | borrow might be used here, when `s` is dropped and runs the destructor for type `S<'_>`
```
